File: uspec/templates/replay.js
// Rebuild the captured uSpec templates as real components in the open file.
//
// WRITES TO FIGMA. Driven by replay.py, which refuses to run without --yes-write.
//
// Constants injected by replay.py: TEMPLATES, EXPECT_FILE, PAGE_NAME, DRY_RUN, ONLY.
//
// Shape of the run:
//   pass 1  create the tree, set content and paint, size everything
//   pass 2  auto-layout, top-down (a child's sizing needs its parent's layoutMode already set)
//   pass 3  sizing / align / grow, top-down
//   pass 4  absolute x/y for children of non-auto-layout parents
// then re-read what was built and hand it back for the diff. The verification is not in this
// file — replay.py compares the returned description against templates.json.

if (figma.root.name !== EXPECT_FILE)
  throw new Error("wrong file open: " + figma.root.name + " (expected " + EXPECT_FILE + ")");

await figma.loadAllPagesAsync();

const wanted = ONLY && ONLY.length ? ONLY : Object.keys(TEMPLATES);

// ---------------------------------------------------------------- fonts

const fontKeys = {};
const collectFonts = (n) => {
  if (n.type === "TEXT") {
    if (n.fontName && n.fontName !== "__MIXED__") fontKeys[n.fontName.family + "|" + n.fontName.style] = 1;
    for (const r of n.runs || []) fontKeys[r.family + "|" + r.style] = 1;
  }
  for (const c of n.children || []) collectFonts(c);
};
for (const w of wanted) collectFonts(TEMPLATES[w]);

const fonts = Object.keys(fontKeys).map((k) => ({ family: k.split("|")[0], style: k.split("|")[1] }));
const fontErrors = [];
for (const f of fonts) {
  try { await figma.loadFontAsync(f); } catch (e) { fontErrors.push(f.family + " " + f.style + ": " + String(e && e.message)); }
}
if (fontErrors.length) throw new Error("fonts unavailable in this file: " + fontErrors.join("; "));

const existing = figma.root.children.filter((p) => p.name === PAGE_NAME).map((p) => p.name);

if (DRY_RUN) {
  const count = (o) => 1 + (o.children || []).reduce((n, c) => n + count(c), 0);
  return {
    dryRun: true,
    file: figma.root.name,
    pageName: PAGE_NAME,
    pageAlreadyExists: existing.length > 0,
    fontsLoaded: fonts,
    wouldCreate: wanted.map((w) => ({ name: w, nodes: count(TEMPLATES[w]) })),
    totalNodes: wanted.reduce((n, w) => n + count(TEMPLATES[w]), 0),
  };
}

// ---------------------------------------------------------------- build

const page = figma.createPage();
page.name = PAGE_NAME;

const SKIP = {
  // derived or set through another channel
  type: 1, name: 1, children: 1, width: 1, height: 1, x: 1, y: 1,
  characters: 1, fontName: 1, runs: 1, vectorPaths: 1, fills: 1, strokes: 1, effects: 1,
  // pass 2 and 3 own these
  layoutMode: 1, layoutWrap: 1, itemSpacing: 1, counterAxisSpacing: 1, itemReverseZIndex: 1,
  paddingTop: 1, paddingRight: 1, paddingBottom: 1, paddingLeft: 1,
  primaryAxisSizingMode: 1, counterAxisSizingMode: 1,
  primaryAxisAlignItems: 1, counterAxisAlignItems: 1,
  layoutSizingHorizontal: 1, layoutSizingVertical: 1,
  layoutAlign: 1, layoutGrow: 1, layoutPositioning: 1,
  // a mixed scalar is described by its per-side siblings
  strokeWeight: 1,
  // read-only in the plugin API
  overflowDirection: 1, autoRename: 1,
};

const set = (node, prop, value) => {
  if (value === "__MIXED__" || value === undefined || value === null) return;
  try { node[prop] = value; } catch (e) { /* property not writable on this node type */ }
};

function build(spec, isRoot) {
  let n;
  if (isRoot) n = figma.createComponent();
  else if (spec.type === "FRAME") n = figma.createFrame();
  else if (spec.type === "TEXT") n = figma.createText();
  else if (spec.type === "VECTOR") n = figma.createVector();
  else throw new Error("unexpected node type in capture: " + spec.type);

  n.name = spec.name;

  if (spec.type === "TEXT") {
    if (spec.fontName && spec.fontName !== "__MIXED__") n.fontName = spec.fontName;
    n.characters = spec.characters || "";
    if (spec.runs) for (const r of spec.runs)
      n.setRangeFontName(r.start, r.end, { family: r.family, style: r.style });
  }
  if (spec.type === "VECTOR" && spec.vectorPaths) n.vectorPaths = spec.vectorPaths;

  if (Array.isArray(spec.fills)) set(n, "fills", spec.fills);
  if (Array.isArray(spec.strokes)) set(n, "strokes", spec.strokes);
  if (Array.isArray(spec.effects)) set(n, "effects", spec.effects);

  for (const [k, v] of Object.entries(spec)) if (!SKIP[k]) set(n, k, v);

  for (const c of spec.children || []) n.appendChild(build(c, false));
  return n;
}

// Auto-layout, parent before child: a child's layoutSizing depends on its parent's layoutMode.
function layout(node, spec) {
  if (spec.layoutMode && spec.layoutMode !== "NONE") {
    node.layoutMode = spec.layoutMode;
    for (const p of ["layoutWrap", "paddingTop", "paddingRight", "paddingBottom", "paddingLeft",
                     "itemSpacing", "counterAxisSpacing", "primaryAxisAlignItems",
                     "counterAxisAlignItems", "itemReverseZIndex"])
      set(node, p, spec[p]);
  }
  const kids = node.children || [];
  (spec.children || []).forEach((cs, i) => { if (kids[i]) layout(kids[i], cs); });
}

function sizing(node, spec) {
  set(node, "layoutPositioning", spec.layoutPositioning);

  // Resize FIRST: resize() writes both axes and silently converts a HUG axis to FIXED, so the
  // captured size has to land before the sizing modes are declared. Getting this backwards is
  // what made the first trial build come back with a FIXED root. See frame.js.
  if (spec.layoutSizingHorizontal === "FIXED" || spec.layoutSizingVertical === "FIXED") {
    try { node.resize(spec.width, spec.height); } catch (e) {}
  }

  set(node, "layoutAlign", spec.layoutAlign);
  set(node, "layoutGrow", spec.layoutGrow);
  set(node, "layoutSizingHorizontal", spec.layoutSizingHorizontal);
  set(node, "layoutSizingVertical", spec.layoutSizingVertical);

  const kids = node.children || [];
  (spec.children || []).forEach((cs, i) => { if (kids[i]) sizing(kids[i], cs); });
}

function place(node, spec) {
  const auto = spec.layoutMode && spec.layoutMode !== "NONE";
  const kids = node.children || [];
  (spec.children || []).forEach((cs, i) => {
    if (!kids[i]) return;
    if (!auto && typeof cs.x === "number") { kids[i].x = cs.x; kids[i].y = cs.y; }
    place(kids[i], cs);
  });
}

const built = [];
let cursorY = 0;
for (const w of wanted) {
  const spec = TEMPLATES[w];
  const comp = build(spec, true);
  page.appendChild(comp);
  try { comp.resize(spec.width, spec.height); } catch (e) {}
  layout(comp, spec);
  sizing(comp, spec);
  place(comp, spec);
  comp.x = 0;
  comp.y = cursorY;
  cursorY += Math.ceil(comp.height) + 200;
  built.push({ name: comp.name, id: comp.id, key: comp.key });
}

// ---------------------------------------------------------------- read back
//
// The same projection capture.js produces, so replay.py can diff shape for shape rather than
// trusting that the writes landed.

const val = (v) => (v === figma.mixed ? "__MIXED__" : v);
const READ = ["visible", "opacity", "blendMode", "layoutMode", "layoutWrap", "layoutAlign",
  "layoutGrow", "layoutPositioning", "primaryAxisAlignItems", "counterAxisAlignItems",
  "layoutSizingHorizontal", "layoutSizingVertical", "itemSpacing", "counterAxisSpacing",
  "paddingTop", "paddingRight", "paddingBottom", "paddingLeft",
  "topLeftRadius", "topRightRadius", "bottomLeftRadius", "bottomRightRadius",
  "strokeAlign", "strokeTopWeight", "strokeRightWeight", "strokeBottomWeight", "strokeLeftWeight",
  "characters", "fontSize", "textAlignHorizontal", "textAlignVertical", "textAutoResize",
  "textCase", "textDecoration", "clipsContent"];

function readBack(n) {
  const o = { name: n.name, type: n.type, width: n.width, height: n.height };
  for (const p of READ) { const v = n[p]; if (v !== undefined && v !== null) o[p] = val(v); }
  if (n.type === "TEXT" && n.fontName && n.fontName !== figma.mixed)
    o.fontName = { family: n.fontName.family, style: n.fontName.style };
  if (Array.isArray(n.fills)) o.fills = JSON.parse(JSON.stringify(n.fills));
  if (n.children) o.children = n.children.map(readBack);
  return o;
}

const result = {};
for (const c of page.children) result[c.name] = readBack(c);

return {
  dryRun: false,
  file: figma.root.name,
  pageName: page.name,
  pageId: page.id,
  built: built,
  readBack: result,
};

File: uspec/templates/replay.py
#!/usr/bin/env python3
"""Rebuild the captured uSpec templates in a Figma file, then verify by reading them back.

    python replay.py --expect-file "<your Figma file>" --dry-run
    python replay.py --expect-file "<your Figma file>" --yes-write

Replay WRITES: it creates a page and ~330 nodes. This workspace requires an explicit go-ahead
for that specific run, so `--yes-write` is mandatory and means exactly that. `--dry-run` is
free and reports what would be created.

Verification is not a screenshot. After building, replay.js reads the new components back out
of the file with the same projection capture.js produces, and this script diffs that against
templates.json property by property. A run that builds but does not match is a failure.
"""

from __future__ import annotations

import argparse
import json
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path

def default_port() -> int:
    """This copy's port, from project.json — the same number the bridge binds."""
    try:
        sys.path.insert(0, str(Path(__file__).resolve().parents[2]))
        import project
        cfg = project.load()
        if cfg:
            return int(cfg["port"])
    except Exception:
        pass
    return 8787


HERE = Path(__file__).resolve().parent

# Properties replay cannot set, or that Figma derives. Excluded from the diff so it reports
# real regressions rather than known physics.
IGNORED = {
    # a mixed scalar is replayed through its per-side siblings
    "strokeWeight",
    # capture records the source position; replay lays the components out on a fresh page
    "x", "y",
    # read-only in the plugin API
    "overflowDirection", "autoRename",
}

# Geometry that auto-layout recomputes. A HUG frame's size is whatever its content makes it,
# so comparing it to the captured pixel is comparing Figma to itself.
GEOMETRY = {"width", "height"}


def post(port: int, code: str, consts: dict, session: str | None, timeout: int) -> dict:
    prelude = "".join("const %s = %s;\n" % (k, json.dumps(v, ensure_ascii=True)) for k, v in consts.items())
    payload = {"code": prelude + code, "timeout": timeout, "want_value": True}
    if session:
        payload["session"] = session
    req = urllib.request.Request(
        "http://localhost:%d/exec" % port,
        data=json.dumps(payload, ensure_ascii=True).encode("utf-8"),
        headers={"Content-Type": "application/json"},
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout + 15) as r:
            body = r.read()
    except urllib.error.HTTPError as e:
        body = e.read()
    except urllib.error.URLError as e:
        sys.exit("bridge unreachable on port %d (%s)" % (port, e))
    resp = json.loads(body)
    if not resp.get("ok"):
        sys.stderr.write("replay failed: %s\n%s\n" % (resp.get("error"), (resp.get("stack") or "")[:2000]))
        sys.exit(1)
    return resp["value"]


def diff(want: dict, got: dict, path: str, out: list) -> None:
    """Compare a captured node against the one that came back, recursively."""
    if got is None:
        out.append((path, "node", "present", "MISSING"))
        return
    for key in ("name", "type"):
        if want.get(key) != got.get(key):
            out.append((path, key, want.get(key), got.get(key)))

    for k, wv in want.items():
        if k in ("children", "fills", "strokes", "effects", "runs", "vectorPaths", "constraints"):
            continue
        if k in IGNORED or k not in got:
            continue
        # `layoutAlign` and `layoutSizing*` are two vocabularies for one state, and the modern
        # one wins: three template nodes report `layoutAlign: STRETCH` together with
        # `layoutSizingHorizontal: FIXED`, which cannot both be acted on. Figma resolves that in
        # favour of the sizing and normalises layoutAlign to INHERIT — so the original and the
        # rebuild lay out identically and only the vestige differs. Forcing STRETCH back would
        # change the real layout, which is worse than the mismatch it silences.
        if k == "layoutAlign" and "layoutSizingHorizontal" in want and "layoutSizingVertical" in want:
            continue
        gv = got[k]
        if k in GEOMETRY:
            # Only meaningful where the node is fixed-size in that axis.
            axis = "layoutSizingHorizontal" if k == "width" else "layoutSizingVertical"
            if want.get(axis) != "FIXED":
                continue
        if isinstance(wv, (int, float)) and isinstance(gv, (int, float)):
            if abs(wv - gv) > 0.01:
                out.append((path, k, round(wv, 2), round(gv, 2)))
        elif wv != gv:
            out.append((path, k, wv, gv))

    # Fills compared as a whole: a wrong colour is the loudest possible regression.
    if isinstance(want.get("fills"), list) and isinstance(got.get("fills"), list):
        if json.dumps(want["fills"], sort_keys=True) != json.dumps(got["fills"], sort_keys=True):
            out.append((path, "fills", "captured", "differs"))

    wc, gc = want.get("children") or [], got.get("children") or []
    if len(wc) != len(gc):
        out.append((path, "childCount", len(wc), len(gc)))
    for i, cw in enumerate(wc):
        diff(cw, gc[i] if i < len(gc) else None, "%s/%s" % (path, cw.get("name")), out)


def main() -> None:
    p = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument("--expect-file", required=True, help="abort unless figma.root.name equals this")
    p.add_argument("--session", help="which connected file to address, when several are open")
    p.add_argument("--page-name", default="uSpec templates")
    p.add_argument("--only", nargs="*", default=[], help="replay only these templates")
    p.add_argument("--port", type=int, default=default_port(),
                   help="bridge port (default: this copy's project.json)")
    p.add_argument("--timeout", type=int, default=300)
    p.add_argument("--dry-run", action="store_true", help="report what would be created, write nothing")
    p.add_argument("--yes-write", action="store_true",
                   help="required to actually build: confirms a go-ahead for this specific run")
    args = p.parse_args()

    if not args.dry_run and not args.yes_write:
        sys.exit(
            "replay creates a page and ~330 nodes in the open Figma file.\n"
            "       This workspace requires an explicit go-ahead for that specific run.\n"
            "       Use --dry-run to see what it would do, or --yes-write once you have one."
        )

    src = json.loads((HERE / "templates.json").read_text(encoding="utf-8"))
    consts = {
        "TEMPLATES": src["templates"],
        "EXPECT_FILE": args.expect_file,
        "PAGE_NAME": args.page_name,
        "DRY_RUN": bool(args.dry_run),
        "ONLY": args.only,
    }

    t0 = time.time()
    v = post(args.port, (HERE / "replay.js").read_text(encoding="utf-8"), consts, args.session, args.timeout)

    if v.get("dryRun"):
        print("DRY RUN — nothing was written to %s" % v["file"])
        print("  page to create: %r%s" % (v["pageName"], "  (A PAGE WITH THIS NAME ALREADY EXISTS)"
                                          if v["pageAlreadyExists"] else ""))
        print("  fonts loaded OK: %s" % ", ".join("%s %s" % (f["family"], f["style"]) for f in v["fontsLoaded"]))
        for w in v["wouldCreate"]:
            print("    %-18s %3d nodes" % (w["name"], w["nodes"]))
        print("  total: %d nodes" % v["totalNodes"])
        print("\n  Re-run with --yes-write once the designer has approved this run.")
        return

    print("built %d templates on page %r of %s in %.1f s"
          % (len(v["built"]), v["pageName"], v["file"], time.time() - t0))
    for b in v["built"]:
        print("    %-18s %-12s key=%s" % (b["name"], b["id"], b["key"]))

    print("\nverifying — read back from the file and diffed against templates.json")
    total = 0
    for name, want in src["templates"].items():
        if args.only and name not in args.only:
            continue
        got = v["readBack"].get(name)
        out: list = []
        diff(want, got, name, out)
        total += len(out)
        print("  %-18s %s" % (name, "OK" if not out else "%d difference(s)" % len(out)))
        for path, prop, wv, gv in out[:12]:
            print("      %-52s %-22s want %r got %r" % (path[-52:], prop, wv, gv))
        if len(out) > 12:
            print("      ... %d more" % (len(out) - 12))

    print("\n%s — %d difference(s) across %d templates"
          % ("VERIFIED" if total == 0 else "MISMATCH", total, len(v["built"])))
    print("Rollback: delete page %s (%s)." % (v["pageName"], v["pageId"]))
    if total:
        sys.exit(1)


if __name__ == "__main__":
    main()
